sql/core/src/main/scala/org/apache/spark/sql/execution/joins/MergeAsOfJoinExec.scala

```diff
@@ -58,23 +69,71 @@ case class MergeAsOfJoinExec(left: SparkPlan, right: SparkPlan, leftOn: Expressi
     }
   }
 
-  protected override def doExecute(): RDD[InternalRow] = {
+  private def match_tolerance(
+      currLeft: (InternalRow, Iterator[InternalRow]),
+      currRight: (InternalRow, Iterator[InternalRow]),
+      duration: Duration,
+      resultProj: InternalRow => InternalRow
+  ): Iterator[InternalRow] = {
+    val joinedRow = new JoinedRow()
+    var rHead = if (currRight._2.hasNext) {
+      currRight._2.next()
+    } else {
+      InternalRow.empty
+    }
+    var rPrev = rHead.copy()
 
-    // basic error checking - both data frames must be sorted by the key
+    currLeft._2.map(lHead => {
+        breakable {
+          while (rHead.getInt(0) <= lHead.getInt(0)) {
+            // TODO make index agnostic and check by type (timestamp)
+            var rHeadCopy = rHead.copy()
+            if (currRight._2.hasNext) {
+              rPrev = rHeadCopy.copy()
+              rHeadCopy = currRight._2.next()
+            } else {
+              break
+            }
+          }
+        }
+        if (rPrev == InternalRow.empty || rPrev.getInt(0) > lHead.getInt(0)) {
+          var dummy = InternalRow(0, 0, 0.0, 0.0) // TODO find better dummy object
+          dummy.setNullAt(1)
+          resultProj(joinedRow(lHead, dummy))
+        } else {
+          resultProj(joinedRow(lHead, rPrev))
+        }
+      }
+    )
+  }
 
-    var numOutputRows: Int = 0
-//    val numOutputRows = longMetric("numOutputRows")
+  protected override def doExecute(): RDD[InternalRow] = {
 
+    val duration = Duration(tolerance)
     val inputSchema = left.output ++ right.output
 
     left.execute().zipPartitions(right.execute()) { (leftIter, rightIter) =>
       val resultProj: InternalRow => InternalRow = UnsafeProjection.create(output, inputSchema)
       if (!leftIter.hasNext || !rightIter.hasNext) {
-        Iterator.empty
+        leftIter
```

Comment:
This doesn't seem right. If the rightIter is empty I think we want to join with empty value here.